blog: Visualizing Complexity (#60)
chickensoft-games/chickensoft_site: content/blog/visualizing-complexity.mdx

  - software architecture
  - architecture
  - game architecture
  - c# game architecture
  - godot project structure
  - c# game structure
  - organize c# project
  - organize godot project
  - Godot
  - C#
---

As a developer, I've worked for many different forms of industries, from fin-tech, to e-commerce, human resources, and even agri-tech. They all had their own style of doing things, some being very organized, while others lacked any form of documentation. Let's face it, documentation is hard, but it doesn't have to be difficult.

{/* truncate */}

One of the things I always look for when being first introduced to a project is some form of relationship diagram. I want to have a big picture of how everything is interconnected. I don't find myself being comfortable with making changes unless I know exactly how this will affect the whole system.

I'd use tools in JetBrains to generate dependency diagrams, and I'd try looking for solutions online, but it would always end up being too complicated, or not exactly giving me the results that I want.

<FancyImage src={"/blogs/visualizing-complexity/rider_diagram.png"} alt="Example of Jetbrains Rider Type Dependency Diagram" />

> While working on my project (Utopia ~ still under development), I use a lot of Chickensoft packages, and I was always fascinated with how the LogicBlocks package generated its own diagrams to show how the state machine functioned. It made me think, how come this sort of thing doesn't exist for the project itself? The whole Godot hierarchy could be represented by a `.puml` diagram, and I wouldn't have to keep running through different scene files trying to remind myself how I set things up again.

Reviewer: ```suggestion
While working on my own game project (_Utopia_ ~ still under development), I have been using a lot of Chickensoft packages. I was always fascinated with how the [LogicBlocks](/docs/logic-blocks) package generates its own diagrams to show how the state machine functioned. It made me think "how come this sort of thing doesn't exist for the project itself? The whole Godot hierarchy could be represented by a `.puml` diagram, and I wouldn't have to keep running through different scene files trying to remind myself how I set things up."
```

> That's when the idea of UMLGenerator came! I quickly came to the Chickensoft community, asking around if my this has ever been done before? Was there any reasons not to? Was there was any interest in a project like this? Sure enough, it looked like this was a good idea and off I went researching on how to accomplish this!

Reviewer: ```suggestion
That's how the idea for UMLGenerator came to me! I asked the Chickensoft community if anyone had done this before or if there there were any reasons not to. I wanted to know if there was any interest in an architecture diagram generator. Sure enough, folks showed interest and I went researching on how to accomplish it!
```

just rewording a bit for clarity

> First of all, If I was going to visualize a hierarchy, we needed to know what children a scene had, and the best way to accomplish this was to check to check the `.tscn` files. Luckily there was already a project: [godot-tscn-parser](https://github.com/MihaMarkic/godot-tscn-parser)

Reviewer: ```suggestion
First of all, If I was going to visualize a hierarchy, I needed to know what children a scene had. The best way to accomplish this was to check to parse the `.tscn` files. Luckily there was already a project to parse Godot scene files: [godot-tscn-parser](https://github.com/MihaMarkic/godot-tscn-parser).
```

If you end up with 3 commas in a sentence, that might mean it's time to split the sentence up 😉 You also can't change pronouns on yourself mid-sentence! ("I was going to...we needed")

> With that out of the way, all that was left was to understand how source generators work. I hadn't built a source generator before, but I sort of had an idea.. Let me start by breaking down the entire process.

Reviewer: ```suggestion
With that out of the way, I simply needed to understand how source generators worked. I hadn't built a source generator before, but I had a vague idea. Let me explain.
```

Not sure if this is what you're trying to say or not, so revise accordingly. 

Expressions like "all that was left was to" are extremely reflective/passive/indirect and should generally be avoided when writing.

> 1. We first need a way to pull all the data. We essentially need all the C# and .tscn files which are core to our game project. The way I accomplished this is by having the user define all the `.tscn` files we want parsed as an AdditionalFile in the `.csproj` files. The `.cs` files are then imported by ensuring that the SyntaxTree file path contains the project directory.

Reviewer: ```suggestion
1. I first needed a way to read all the data. I essentially need all the C# and .tscn files which are core to a game project. I accomplished this by having the user define all the `.tscn` files we want parsed as an `AdditionalFile` in the `.csproj` file. The `.cs` files are then imported by ensuring that the `SyntaxTree` file path contains the project directory.
```

You can't keep changing pronouns! You either pick a "we"-based article or an "I"-based article. There are times when you can slip "I" into a "we"-based article when you are actually talking about yourself as an individual, but this isn't that time. You need to pick a point of view and stick to it. 😛

<Callout type="info">

>     I did consider also having all the `.tscn` files in the project directory be imported, but then I ran into some parsing problems with some godot addons, so ultimately I decided against it

Reviewer: ```suggestion
    I did consider importing the `.tscn` files in the project directory, but I ran into some parsing problems with godot addons and ultimately decided against it.
```

</Callout>

> 2. Afterwards we link everything together as parent child relationships. `.cs` files with the \[ClassDiagram] attribute are used as the root of the hierarchy. `.tscn` files which have a root `.cs` script are paired together. Then everything is given a assigned a relationship based on the hierarchy of the `.tscn` or the properties that the `.cs` file has.

Reviewer: ```suggestion
2. Afterwards I link everything together via parent-child relationships. The `.cs` files with \[ClassDiagram] attributes are used as the root of the hierarchies. Each `.tscn` file which has a root `.cs` script are paired together. Then, everything is given an assigned relationship based on the hierarchy of the scene file or the properties in its script file (if it has one).
```

> 3. Once everything is linked together it's just a matter of recursively going down the hierarchy and returning the diagram output string of each child and concatenating everything together.

Reviewer: ```suggestion
3. Once everything is linked together, it's just a matter of recursively going down the hierarchy and returning the diagram output string of each child and concatenating everything together.
```

> Example:
> Given the following project hierarchy

Reviewer: ```suggestion
For example:
```

```mermaid
classDiagram

  Game --> GameRepo
  Game --> GameLogic

  class Game {
    obj GameRepo
    obj GameLogic
    void StartGame()
    void LoadGame()
    void SaveGame()
  }

  class GameRepo {

  }

  class GameLogic {

  }

```

> GameRepo and GameLogic would be linked as children to the Game class.

Reviewer: ```suggestion
`GameRepo` and `GameLogic` would be linked as children to the `Game` class.
```

> Then we would call the method to generate the diagram for the Game class, which would then call the respective methods to generate a diagram for the GameRepo and GameLogic classes.

Reviewer: ```suggestion
The generator would then call the method to generate the diagram for the `Game` class, which in turn calls the respective methods to generate a diagram for the `GameRepo` and `GameLogic` classes.
```

> GameRepo and GameLogic would return their respective diagrams, be concatenated with each other, and then be nested under the diagram for the Game class.

Reviewer: ```suggestion
`GameRepo` and `GameLogic` would return their respective diagrams, be concatenated with each other, and then be nested under the diagram for the `Game` class.
```

<details>
<summary><code>Game.g.puml</code> Example</summary>
```PlantUML
package Game-Scene [[../../TestCases/BaseProject_Rider/Game.tscn]] {

class Game {

[[../../Game/Game.cs ScriptFile]]

--

[[../../Game/Game.cs:22 GameLogic]] - [[../../TestCases/BaseProject_Rider/Logic/GameLogic.cs Script]]

[[../../Game/Game.cs:21 GameRepo]] - [[../../TestCases/BaseProject_Rider/Domain/GameRepo.cs Script]]

}

class GameRepo {

[[../../TestCases/BaseProject_Rider/Domain/GameRepo.cs ScriptFile]]

}

class GameLogic {

[[../../TestCases/BaseProject_Rider/Logic/GameLogic.cs ScriptFile]]

}

Game::GameRepo --> GameRepo

Game::GameLogic --> GameLogic

}
```
</details>

> As your project grows, so does your diagram. What starts off as a basic visualization becomes a full-blown map of your application, one that evolves with every new class, every scene change, every feature.

Reviewer: ```suggestion
As your project grows, so too does your diagram. What starts off as a basic visualization becomes a full-blown map of your application that evolves with every new class, scene change, and feature.
```

> We’ve added logic to simplify output by avoiding rendering child nodes unless they also have children, keeping things clean and readable. But the point is: **you no longer have to memorize your codebase’s architecture.** It’s right there, clickable and up-to-date.

Reviewer: ```suggestion
We’ve added logic to simplify output by avoiding rendering child nodes unless they also have children, keeping things clean and readable. **You no longer have to memorize your codebase’s architecture.** It’s right there, clickable and up-to-date.
```

> UMLGenerator is still in active development, and there’s plenty more I want to add like:

Reviewer: ```suggestion
Note that UMLGenerator is still in active development — there’s plenty more I want to add:
```

> - Find UseTransient/UseSingleton/UseScoped to draw relationships

Reviewer: ```suggestion
- Find `UseTransient`/`UseSingleton`/`UseScoped` to draw relationships
```

- Look for lists of components and draw one to many relationships
- Create Godot plugin for non-C# projects

> But even in its current form, it’s already made my life easier — and if you’re working on a Godot + C# project, I think it could do the same for you.

Reviewer: ```suggestion
Even in its current form, UMLGenerator has already made my life easier. If you’re working on a Godot + C# project, I think it could do the same for you.
```
